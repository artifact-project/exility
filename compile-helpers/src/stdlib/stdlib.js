"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var R_HTML_ENTITES = /[&<>"]/;
var R_HTML_ENTITY_AMP = /&/g;
var R_HTML_ENTITY_LT = /</g;
var R_HTML_ENTITY_GT = />/g;
var R_HTML_ENTITY_QUOTE = /"/g;
function NOOP() { }
exports.default = {
    NIL: null,
    R_HTML_ENTITES: R_HTML_ENTITES,
    R_HTML_ENTITY_AMP: R_HTML_ENTITY_AMP,
    R_HTML_ENTITY_LT: R_HTML_ENTITY_LT,
    R_HTML_ENTITY_GT: R_HTML_ENTITY_GT,
    R_HTML_ENTITY_QUOTE: R_HTML_ENTITY_QUOTE,
    NOOP: NOOP,
    TO_STRING: function (value) {
        return value == null ? '' : value + '';
    },
    RETURN_EMPTY_STRING: function () {
        return '';
    },
    HTML_ENCODE: function (value) {
        if (value == null) {
            return '';
        }
        if (R_HTML_ENTITES.test(value)) {
            if (value.indexOf('&') !== -1) {
                value = value.replace(R_HTML_ENTITY_AMP, '&amp;');
            }
            if (value.indexOf('<') !== -1) {
                value = value.replace(R_HTML_ENTITY_LT, '&lt;');
            }
            if (value.indexOf('>') !== -1) {
                value = value.replace(R_HTML_ENTITY_GT, '&gt;');
            }
            if (value.indexOf('"') !== -1) {
                value = value.replace(R_HTML_ENTITY_QUOTE, '&quot;');
            }
            return value;
        }
        return value.toString();
    },
    EACH: function (data, callback) {
        if (data != null) {
            if (data.forEach) {
                var length_1 = data.length;
                for (var i = 0; i < length_1; i++) {
                    callback(data[i], i);
                }
            }
            else {
                for (var key in data) {
                    if (data.hasOwnProperty(key)) {
                        callback(data[key], key);
                    }
                }
            }
        }
    },
    SLOT: function (parent, slots, __super__, name, fn) {
        if (slots && slots.hasOwnProperty(name)) {
            slots[name](parent, __super__);
        }
        else {
            fn();
        }
    },
    CLONE_OBJECT: function (target) {
        var cloned = {};
        if (target != null) {
            for (var key in target) {
                cloned[key] = target[key];
            }
        }
        return cloned;
    }
};
