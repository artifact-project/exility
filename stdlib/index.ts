import core from './src/core/core';
import dom from './src/dom/dom';

export {
	core,
	dom,
};
