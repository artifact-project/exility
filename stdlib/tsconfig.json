{
  "compilerOptions": {
    "module": "commonjs",
    "target": "es5",
    "declaration": true,
    "moduleResolution": "node",
    "experimentalDecorators": true,
    "importHelpers": true,
    "noEmitHelpers": true,
    "lib": [
      "dom",
      "es5",
      "es2015.promise",
      "es2016.array.include"
    ]
  },
  "exclude": [
    "node_modules"
  ]
}
