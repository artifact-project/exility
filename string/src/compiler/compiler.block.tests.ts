import {core as stdlib} from '@exility/stdlib';
import Block from '@exility/block';
import compilerFactory from './compiler';

function fromString(input: string, scope?: string[], blocks = {}) {
	if (blocks) {
		scope.push('__blocks__');
	}

	const compiler = compilerFactory({
		debug: true,
		scope,
		blocks: Object.keys(blocks),
	});

	const factory = compiler(input);
	const template = factory({stdlib});

	template.factory = factory;

	return template;
}

it('Icon', () => {
	const template = fromString(`
		Icon = [name]
			i.icon-\${name}

		Icon[name=\${x}]
	`, ['x']);

	expect(template.factory).toMatchSnapshot();
	expect(template({x: 'foo'})).toBe('<i class="icon-foo"></i>');
});

it('Icon (short)', () => {
	const template = fromString(`
		Icon = [name] > i.icon-\${name}

		Icon[name=\${x}]
	`, ['x']);

	expect(template.factory).toMatchSnapshot();
	expect(template({x: 'foo'})).toBe('<i class="icon-foo"></i>');
});

it('XIf', () => {
	const template = fromString(`
		XIf = [expr] > if (expr) > b
		
		XIf[expr=\${val}]
	`, ['val']);

	expect(template({val: false})).toBe('');
	expect(template({val: true})).toBe('<b></b>');
});

it('Btn -> Sync', () => {
	const __blocks__ = {'Btn': {template: 'button | ${attrs.value}'}};
	const template = fromString('Btn[value=${val}]', ['val'], __blocks__);

	expect(template({val: 'Wow', __blocks__})).toBe('<button>Wow</button>');
});

it('Btn -> Async', async () => {
	const __blocks__ = {'Btn': () => {}};
	const template = fromString('Btn[value=${val}]', ['val'], __blocks__);

	expect(template({val: 'Wow', __blocks__})).toBe(
		'<div data-block=\"Btn\" class=\"block-dummy block-dummy-loading\"></div>'
	);
});

it('Pseudo Elements', () => {
	const __blocks__ = {'Pseudo': {template: 'div > ::children | ${attrs.value}'}};
	const template = fromString(`Pseudo[value=\${text}]`, ['text'], __blocks__);

	expect(template({text: 'Wow', __blocks__})).toBe('<div>Wow</div>');
	expect(template.factory).toMatchSnapshot();
});

it('Pseudo Elements: override default', () => {
	const __blocks__ = {'Pseudo': {template: 'div > ::children | fail'}};
	const template = fromString(`Pseudo | OK`, [], __blocks__);

	expect(template({__blocks__})).toBe('<div>OK</div>');
	expect(template.factory).toMatchSnapshot();
});

it('Pseudo Elements: override by name', () => {
	const __blocks__ = {'Pseudo': {template: 'div > ::slotik | fail'}};
	const template = fromString(`Pseudo > ::slotik | OK`, [], __blocks__);

	expect(template({__blocks__})).toBe('<div>OK</div>');
	expect(template.factory).toMatchSnapshot();
});

it('Pseudo Elements: super', () => {
	const Pseudo = Block.classify({template: 'div > ::children | Original:${attrs.value}'});
	const __blocks__ = {Pseudo};
	const template = fromString(
		'Pseudo[value=${x}] > ::children\n  | ${text}:\n  ::super.children',
		['x', 'text'],
		__blocks__,
	);

	expect(template.factory).toMatchSnapshot();
	expect(Pseudo.prototype['__template__']).toMatchSnapshot();
	expect(template({x: 1, text: 'OK', __blocks__})).toBe('<div>OK:Original:1</div>');
});
