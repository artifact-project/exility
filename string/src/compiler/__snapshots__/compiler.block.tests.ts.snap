// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon (short) 1`] = `
function templateFactory(__DEPS__) {
	"use strict";
	var __STDLIB = __DEPS__.stdlib;
	var __STDLIB_HTML_ENCODE = __STDLIB.HTML_ENCODE;
	function Icon(attrs, __slots) {
		var __super = {attrs: attrs},
		name = attrs.name
		return "<i class=\\"icon-" + __STDLIB_HTML_ENCODE(name) + "\\"></i>";
	}

	return function compiledTemplate(__SCOPE__, __OPTIONS__) {
		__SCOPE__ = __SCOPE__ || {};
		var x = __SCOPE__.x;
		var __blocks__ = __SCOPE__.__blocks__;

		return Icon({name: (x)});
	}
}
`;

exports[`Icon 1`] = `
function templateFactory(__DEPS__) {
	"use strict";
	var __STDLIB = __DEPS__.stdlib;
	var __STDLIB_HTML_ENCODE = __STDLIB.HTML_ENCODE;
	function Icon(attrs, __slots) {
		var __super = {attrs: attrs},
		name = attrs.name
		return "<i class=\\"icon-" + __STDLIB_HTML_ENCODE(name) + "\\"></i>";
	}

	return function compiledTemplate(__SCOPE__, __OPTIONS__) {
		__SCOPE__ = __SCOPE__ || {};
		var x = __SCOPE__.x;
		var __blocks__ = __SCOPE__.__blocks__;

		return Icon({name: (x)});
	}
}
`;

exports[`Pseudo Elements 1`] = `
function templateFactory() {
	"use strict";
	function __BLOCK_INIT(blocks, name) {
		var XBlock = blocks[name];

		if (XBlock.length !== 0) {
			blocks[name] = __COMPILER__.compileBlock(blocks[name]);
		}
	}

	function __BLOCK_RENDER(blocks, name, attrs, slots) {
		var XBlock = blocks[name];

		if (XBlock.length) {
			var block = new XBlock(attrs, {slots: slots});
			return block.__view__;
		} else {
			return '<div data-block="' + name + '" class="block-dummy block-dummy-loading"></div>';
		}
	}

	return function compiledTemplate(__SCOPE__, __OPTIONS__) {
		__SCOPE__ = __SCOPE__ || {};
		var text = __SCOPE__.text;
		var __blocks__ = __SCOPE__.__blocks__;

		__BLOCK_INIT(__blocks__, "Pseudo");
		return __BLOCK_RENDER(__blocks__, "Pseudo", {value: (text)}, null);
	}
}
`;

exports[`Pseudo Elements: override by name 1`] = `
function templateFactory() {
	"use strict";
	function __BLOCK_INIT(blocks, name) {
		var XBlock = blocks[name];

		if (XBlock.length !== 0) {
			blocks[name] = __COMPILER__.compileBlock(blocks[name]);
		}
	}

	function __BLOCK_RENDER(blocks, name, attrs, slots) {
		var XBlock = blocks[name];

		if (XBlock.length) {
			var block = new XBlock(attrs, {slots: slots});
			return block.__view__;
		} else {
			return '<div data-block="' + name + '" class="block-dummy block-dummy-loading"></div>';
		}
	}

	return function compiledTemplate(__SCOPE__, __OPTIONS__) {
		__SCOPE__ = __SCOPE__ || {};
		var __blocks__ = __SCOPE__.__blocks__;

		__BLOCK_INIT(__blocks__, "Pseudo");
		var __ROOT = __BLOCK_RENDER(__blocks__, "Pseudo", {}, {slotik: function __SLOT_slotik(__super__) {
			return "OK";
			}});

			return __ROOT
		}
	}
`;

exports[`Pseudo Elements: override default 1`] = `
function templateFactory() {
	"use strict";
	function __BLOCK_INIT(blocks, name) {
		var XBlock = blocks[name];

		if (XBlock.length !== 0) {
			blocks[name] = __COMPILER__.compileBlock(blocks[name]);
		}
	}

	function __BLOCK_RENDER(blocks, name, attrs, slots) {
		var XBlock = blocks[name];

		if (XBlock.length) {
			var block = new XBlock(attrs, {slots: slots});
			return block.__view__;
		} else {
			return '<div data-block="' + name + '" class="block-dummy block-dummy-loading"></div>';
		}
	}

	return function compiledTemplate(__SCOPE__, __OPTIONS__) {
		__SCOPE__ = __SCOPE__ || {};
		var __blocks__ = __SCOPE__.__blocks__;

		__BLOCK_INIT(__blocks__, "Pseudo");
		var __ROOT = __BLOCK_RENDER(__blocks__, "Pseudo", {}, {children: function __SLOT_children(__super__) {
			return "OK";
			}});

			return __ROOT
		}
	}
`;

exports[`Pseudo Elements: super 1`] = `
function templateFactory(__DEPS__) {
	"use strict";
	var __STDLIB = __DEPS__.stdlib;
	var __STDLIB_HTML_ENCODE = __STDLIB.HTML_ENCODE;
	function __BLOCK_INIT(blocks, name) {
		var XBlock = blocks[name];

		if (XBlock.length !== 0) {
			blocks[name] = __COMPILER__.compileBlock(blocks[name]);
		}
	}

	function __BLOCK_RENDER(blocks, name, attrs, slots) {
		var XBlock = blocks[name];

		if (XBlock.length) {
			var block = new XBlock(attrs, {slots: slots});
			return block.__view__;
		} else {
			return '<div data-block="' + name + '" class="block-dummy block-dummy-loading"></div>';
		}
	}

	return function compiledTemplate(__SCOPE__, __OPTIONS__) {
		__SCOPE__ = __SCOPE__ || {};
		var x = __SCOPE__.x;
		var text = __SCOPE__.text;
		var __blocks__ = __SCOPE__.__blocks__;

		__BLOCK_INIT(__blocks__, "Pseudo");
		var __ROOT = __BLOCK_RENDER(__blocks__, "Pseudo", {value: (x)}, {children: function __SLOT_children(__super__) {
			return __STDLIB_HTML_ENCODE(text) + ":" + __super__["children"]();
			}});

			return __ROOT
		}
	}
`;

exports[`Pseudo Elements: super 2`] = `undefined`;
